presto-main/src/test/java/com/facebook/presto/cost/TestFilterStatsCalculator.java

```diff
@@ -158,6 +158,31 @@ public void testBooleanLiteralStats()
                 .symbolStats("rightOpen", SymbolStatsAssertion::empty)
                 .symbolStats("emptyRange", SymbolStatsAssertion::empty)
                 .symbolStats("unknownRange", SymbolStatsAssertion::empty);
+
+        // null will be evaluated to false
+        assertExpression("cast(null as boolean)")
+                .outputRowsCount(0.0)
+                .symbolStats("x", SymbolStatsAssertion::empty)
+                .symbolStats("y", SymbolStatsAssertion::empty)
+                .symbolStats("z", SymbolStatsAssertion::empty)
+                .symbolStats("leftOpen", SymbolStatsAssertion::empty)
+                .symbolStats("rightOpen", SymbolStatsAssertion::empty)
+                .symbolStats("emptyRange", SymbolStatsAssertion::empty)
+                .symbolStats("unknownRange", SymbolStatsAssertion::empty);
+
+        // more complicated expression containing a null will return unknown; the interpreter does not return null unless all operands are null
+        assertExpression("cast(null as boolean) AND sin(x) > x").outputRowsCountUnknown();
```

Comment:
Why unknown? It should return `outputRowsCount(0.0)`. @rschlussel ?

Reply:
Because the way we interpret `NULL` in a logical expression is kinda weird:
```
presto:di> select null and true;
 _col0
-------
 NULL
(1 row)

Query 20190425_221039_96786_k7smy, FINISHED, 1 node
Splits: 5 total, 5 done (100.00%)
0:00 [0 rows, 0B] [0 rows/s, 0B/s]

presto:di> select null or true;
 _col0
-------
 true
(1 row)

Query 20190425_221045_96875_k7smy, FINISHED, 1 node
Splits: 5 total, 5 done (100.00%)
0:00 [0 rows, 0B] [0 rows/s, 0B/s]

presto:di> select null and false;
 _col0
-------
 false
(1 row)

Query 20190425_221217_98586_k7smy, FINISHED, 1 node
Splits: 5 total, 5 done (100.00%)
0:00 [0 rows, 0B] [0 rows/s, 0B/s]

presto:di> select null or false;
 _col0
-------
 NULL
(1 row)
```
So, if we see a null in a logical expression, we don't quite know if the output is true, false, or null.

Reply:
@highker This is `FilterStatCalculator`. 

Projection stats are calculated by `ScalarStatsCalculator`:

https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java#L349
https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java#L153

For a filter

```
SELECT *
FROM table
WHERE null
```

always return 0 rows.

Reply:
yeah, Andrii's right for final filter stats the only thing we care about is true and not true, so null is the same as false.  The only trick with null is to make sure negation is handled correctly not(null) is also false.